src/Microsoft.AspNetCore.Authentication/Events/AuthenticateResultContext.cs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Claims;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Authentication
+{
+    /// <summary>
+    /// Base context for authentication.
+    /// </summary>
+    public abstract class AuthenticateResultContext<TOptions> : BaseContext<TOptions> where TOptions : AuthenticationSchemeOptions
+    {
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <param name="scheme">The authentication scheme.</param>
+        /// <param name="options">The authentication options associated with the scheme.</param>
+        protected AuthenticateResultContext(
+            HttpContext context,
+            AuthenticationScheme scheme,
+            TOptions options)
+            : base(context, scheme, options)
+        {
+        }
+
+        /// <summary>
+        /// Gets or set the <see cref="AuthenticationTicket"/> containing
+        /// the user principal and the authentication properties.
+        /// </summary>
+        public AuthenticationTicket Ticket { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="ClaimsPrincipal"/> containing the user claims.
+        /// </summary>
+        public ClaimsPrincipal Principal => Ticket?.Principal;
+
+        /// <summary>
+        /// Gets or sets the <see cref="AuthenticationProperties"/>.
+        /// </summary>
+        public override AuthenticationProperties Properties
```

Comment:
I took a look and here's what I found:

This pattern (setting both the ticket and the properties) is used for two events: `TokenValidated` and `AuthorizationCodeReceived`.

For `TokenValidated`, I believe we could avoid setting `Properties` explicitly in this case because we always have a ticket (created from the identity token extracted from the authorization response or from the token response).

For `AuthorizationCodeReceived`, it's much more complicated: we always have an `AuthenticationProperties` instance (resolved from the `state` very early in the request handling), but not necessarily a ticket (e.g if the response doesn't contain an identity token, like in the code flow). Setting both `Ticket` and `Properties` sounds like the only option.